### ISSUE TITLE
Performance problem with pagination.

### ISSUE THREAD

**Issue by billk97** [Nooul/spring-boot-rest-api-helpers#21]
The library seems to have a performance problem with large tables.
While the actual query is fast by it self, in order to serve pageable data it has to know the total amount of results.
To achieve that it performs the bellow query, which is extremely slow 36s in the below example.
```sql
select distinct count(distinct message0_.id) as col_0_0_ from message message0_ inner join application_user applicatio1_ on message0_.created_by_id=applicatio1_.id where applicatio1_.id=5
```
The problem seems to be in the `distinct`  which is added to the query. 
By simply removing the `distinct` as shown bellow the performance improvement was significant.
Total table rows: 2.300.000
time with distinct: 32sec
time **without** distinct: 0,9sec
  
```sql
select count(message0_.id) as col_0_0_ from message message0_ inner join application_user applicatio1_ on message0_.created_by_id=applicatio1_.id where applicatio1_.id=5
```
Sample image of the transactions as shown in our monitoring tool. 
![image](https://user-images.githubusercontent.com/33226904/155290583-bd8f9041-6a5c-43af-9f07-1ca4f1c6d53c.png)

**Comment by zifnab87** [comment]
yeah but is the count correct for all the queries if distinct is removed?
also the filtering may be done on tables that dont have proper indices? I
will have a look when possible

**Comment by billk97** [comment]
You are right, just removing the `distinct` will introduce duplicates in some scenarios. This can be easily demonstrated in your tests. However, depending on the use case the data could already be distinct for example a UUID or an incremental id in the DB. I added a pull request 
that will allow disabling the distinct in such scenarios. Please review it and comment if needed.  
Examples: 
normal query
`GET /actors?filter={movies: 1, firstName: John}`
disabling distinct query
`GET /actors?filter={movies: 1, firstName: John, disableDistinct: true}`
Do you think there is a better way to disable the distinct?

**Comment by zifnab87** [comment]
if the indices are right the mysql optimizer should have the same
performance with or without distinct so you may need to look at the
combination of indices. Also if it is uniqueness on uuid it will be way
slower than ints since comparison is so much slower to see for uniqueness.
I will have a look in the pull request. Thanks

Στις Τετ 23 Φεβ 2022 στις 1:48 μ.μ., ο/η Vasilis Konstantinou <
***@***.***> έγραψε:

**Comment by zifnab87** [comment]
closed as changes were merged by #22
